### PR TITLE
[SPARK-45093][CONNECT][PYTHON] Properly support error handling and conversion for AddArtifactHandler

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
This patch improves the error handling when errors are happening in the `AddArtifact` path. In particular, the `AddArtifactHandler` would not properly return exceptions but all exceptions would end up yielding `UNKNOWN` errors. This patch makes sure we wrap the errors in the add artifact path the same way as we're wrapping the errors in the normal query execution path.

In addition, it adds tests and verification for this behavior in Scala and in Python.

### Why are the changes needed?
Stability

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT for Scala and Python

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by grundprinzip** [comment]
@hvanhovell @HyukjinKwon PTAL

**Comment by HyukjinKwon** [comment]
cc @cdkrot @heyihong mind taking a look please?

**Review by HyukjinKwon — APPROVED** [review]

**Review by heyihong — APPROVED** [review]
LGTM

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala [hunk 1]
```diff
@@ -128,7 +136,7 @@ class SparkConnectAddArtifactsHandler(val responseObserver: StreamObserver[AddAr
       responseObserver.onNext(builder.build())
       responseObserver.onCompleted()
     } catch {
-      case e: StatusRuntimeException => onError(e)
+      ErrorUtils.handleNonQueryError("addArtifact - onNext", responseObserver, Some(this.holder))
```
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:139]: nit: should this op name be "addArtifact - onCompleted"?
nit2: the RPC is named "addArtifacts"; should we use opname "addArtifacts"?
nit3: maybe "addArtifacts.onNext" / "addArtifacts.onCompleted"? I somehow find the " - " separator weird.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala [hunk 2]
```diff
@@ -213,6 +221,30 @@ private[connect] object ErrorUtils extends Logging {
       .exists(_.toString.contains("org.apache.spark.sql.execution.python"))
   }
 
+  def handleNonQueryError[V](
+      opName: String,
+      observer: StreamObserver[V],
+      sessionHolderOpt: Option[SessionHolder],
+      callback: Option[() => Unit] = None): PartialFunction[Throwable, Unit] = {
```
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala:228]: does this need to be a separate function from `handleError`, or could the "callback" be called into the previous function?
At a glance `handleError` seems to be a compatible superset of functionality.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala [hunk 3]
```diff
@@ -213,6 +221,30 @@ private[connect] object ErrorUtils extends Logging {
...
+    val partial: PartialFunction[Throwable, Throwable] = {
+      case e: Throwable if e.isInstanceOf[SparkThrowable] || NonFatal.apply(e) =>
+        StatusProto.toStatusRuntimeException(buildStatusFromThrowable(e, sessionHolderOpt))
+      case e: StatusRuntimeException => e
```
Comment (juliuszsompolski) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala:232]: this case is not present in the query `handleError`, why it's a case needed here?

Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala:232]: In the original query error case, I was very certain that the error conversion does not happen inside the try block. Let me research why I added this (or if I was just cautious)

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala (hunk 1 above)
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:139]: You could have saved yourself a large number of characters by just proposing the last one. Changing.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala [hunk 4]
```diff
@@ -213,6 +221,30 @@ private[connect] object ErrorUtils extends Logging {
       .exists(_.toString.contains("org.apache.spark.sql.execution.python"))
   }
 
+  def handleNonQueryError[V](
+      opName: String,
+      observer: StreamObserver[V],
+      sessionHolderOpt: Option[SessionHolder],
```
[the remark this answers is not in the document]
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala:227]: This function was removed.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala [hunk 5]
```diff
@@ -78,6 +78,22 @@ class SparkConnectAddArtifactsHandler(val responseObserver: StreamObserver[AddAr
     } else {
       throw new UnsupportedOperationException(s"Unsupported data transfer request: $req")
     }
+  } catch {
+    ErrorUtils.handleError(
```
Comment (heyihong) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:82]: We need to include cleanUpStagedArtifacts callback?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala [hunk 6]
```diff
@@ -230,7 +249,8 @@ private[connect] object ErrorUtils extends Logging {
       userId: String,
       sessionId: String,
       events: Option[ExecuteEventsManager] = None,
-      isInterrupted: Boolean = false): PartialFunction[Throwable, Unit] = {
+      isInterrupted: Boolean = false,
+      callback: Option[() => Unit] = None): PartialFunction[Throwable, Unit] = {
```
Comment (heyihong) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala:253]: callback is unused?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala [hunk 7]
```diff
@@ -128,7 +144,10 @@ class SparkConnectAddArtifactsHandler(val responseObserver: StreamObserver[AddAr
       responseObserver.onNext(builder.build())
       responseObserver.onCompleted()
     } catch {
-      case e: StatusRuntimeException => onError(e)
+      ErrorUtils.handleNonQueryError(
```
Comment (heyihong) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectAddArtifactsHandler.scala:147]: ErrorUtils.handleError?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala (hunk 6 above)
Comment (grundprinzip) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/utils/ErrorUtils.scala:253]: yeah, too early push

**Pull request closed without merging**